library/download-manager/src/main/java/org/cru/godtools/download/manager/GodToolsDownloadManager.kt

```diff
@@ -0,0 +1,363 @@
+import androidx.annotation.RestrictTo
+import androidx.annotation.VisibleForTesting
+import androidx.annotation.WorkerThread
+import androidx.collection.ArrayMap
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import com.google.common.io.CountingInputStream
+import java.io.FileNotFoundException
+import java.io.IOException
+import java.io.InputStream
+import java.util.Locale
+import java.util.zip.ZipInputStream
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.actor
+import kotlinx.coroutines.channels.receiveOrNull
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.withTimeoutOrNull
+import org.ccci.gto.android.common.base.TimeConstants.HOUR_IN_MS
+import org.ccci.gto.android.common.base.TimeConstants.MIN_IN_MS
+import org.ccci.gto.android.common.db.Expression
+import org.ccci.gto.android.common.db.Query
+import org.ccci.gto.android.common.db.find
+import org.ccci.gto.android.common.db.get
+import org.ccci.gto.android.common.kotlin.coroutines.MutexMap
+import org.ccci.gto.android.common.kotlin.coroutines.ReadWriteMutex
+import org.ccci.gto.android.common.kotlin.coroutines.withLock
+import org.ccci.gto.android.common.util.ThreadUtils
+import org.cru.godtools.api.AttachmentsApi
+import org.cru.godtools.base.FileManager
+import org.cru.godtools.model.Attachment
+import org.cru.godtools.model.Language
+import org.cru.godtools.model.LocalFile
+import org.cru.godtools.model.Tool
+import org.cru.godtools.model.Translation
+import org.cru.godtools.model.TranslationFile
+import org.cru.godtools.model.TranslationKey
+import org.cru.godtools.model.event.AttachmentUpdateEvent
+import org.cru.godtools.model.event.LanguageUpdateEvent
+import org.cru.godtools.model.event.ToolUpdateEvent
+import org.cru.godtools.model.event.TranslationUpdateEvent
+import org.greenrobot.eventbus.EventBus
+import org.keynote.godtools.android.db.Contract.AttachmentTable
+import org.keynote.godtools.android.db.Contract.LanguageTable
+import org.keynote.godtools.android.db.Contract.LocalFileTable
+import org.keynote.godtools.android.db.Contract.ToolTable
+import org.keynote.godtools.android.db.Contract.TranslationFileTable
+import org.keynote.godtools.android.db.Contract.TranslationTable
+import org.keynote.godtools.android.db.GodToolsDao
+
+private const val TAG = "GodToolsDownloadManager"
+
+@VisibleForTesting
+internal const val CLEANUP_DELAY = HOUR_IN_MS
+@VisibleForTesting
+internal const val CLEANUP_DELAY_INITIAL = MIN_IN_MS
+
+open class KotlinGodToolsDownloadManager @VisibleForTesting internal constructor(
```

Comment:
I've looked over this and I think I have an idea of what's going on, just not sure how it's all working if that makes sense haha.